### PR TITLE
Transformer predictor for JEPA

### LINKED ISSUE #1587: Implement transformer based predictors for JEPA
### Describe the task. It can be a feature, documentation, etc.

We need to implement a transformer based predictor for JEPA, following the literature (see https://www.overleaf.com/project/68dd158b1e327e168d9ceebd)

We will use the existing attention blocks and MLP layers for this. This resembles the forecasting engine architecture. A quick look revelas that it is not very straight forward to refactor these two to a single class, so first we will implement it as a seperate class.

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [x] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation

### PR DESCRIPTION
**Pull request by kctezcan**
## Issue Number

Closes #1587 

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [x] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

**Comment by kctezcan** [comment]
Some design choices to discuss:
1) the latent space has the default dimensions of 12288x2048. The optimal predictor dimension is 384. We use a linear layer without bias at the entrance of the predictor to map 2048->384 and as the last layer another linear map without bias to map 384->2048. These dimensions are all config parameters.
2) We hardcoded 4 as the  MLP factor (factor for the hidden dimension in the MLP layers) as this is the optimal value from the literature.
3) we have introduced new parameters only in the jepa config: `pred_xxx`

File: config/default_config_jepa.yml [hunk 1]
```diff
@@ -97,6 +103,7 @@ ema_halflife_in_thousands: 1e-3
 # ### Example validation and training config for student-teacher with JEPA
 validation_config:
   losses:
+    LossPhysical: {weight: 0.0, loss_fcts: [['mse', 0.8], ['mae', 0.2]]}
```
Comment (sophie-xhonneux) [inline config/default_config_jepa.yml]: can we remove this if instead of setting the weight to 0?

File: config/default_config_jepa.yml [hunk 2]
```diff
@@ -106,6 +113,7 @@ training_config:
   training_mode: "student_teacher"  # "masking", "student_teacher", "forecast"
   target_and_aux_calc: "EMATeacher"
   losses :  
+    LossPhysical: {weight: 0.0, loss_fcts: [['mse', 0.8], ['mae', 0.2]]}
```
Comment (sophie-xhonneux) [inline config/default_config_jepa.yml]: Same here

**Review by sophie-xhonneux — APPROVED** [review]
Broadly, I am happy to approve the comments are addressed.

But I would like to talk about whether the global assimilation engine makes sense to be a part of the encoder or if for JEPA we should instead use the AggregationEngine (Global transformer without the masked tokens)

File: config/default_config_jepa.yml (hunk 1 above)
Comment (kctezcan) [reply config/default_config_jepa.yml]: this file is removed with the new configs

File: config/default_config_jepa.yml (hunk 2 above)
Comment (kctezcan) [reply config/default_config_jepa.yml]: this file is removed with the new configs

**Comment by kctezcan** [comment]
I renamed the new parameters related to the JEPA predictor from `pred_` to `sslpred_` because I realized there were already parameters called `pred_` for the decoder (i.e. prediction heads)

**Review by sophie-xhonneux — APPROVED** [review]

File: config/config_physical_jepa.yml [hunk 3]
```diff
@@ -65,6 +65,12 @@ fe_impute_latent_noise_std: 0.0  # 1e-4
 forecast_att_dense_rate: 1.0
 with_step_conditioning: True # False
 
+sslpred_num_blocks: 12
```
Comment (clessig) [inline config/config_physical_jepa.yml:68]: Where would this go into the new config structure?

File: src/weathergen/model/engines.py [hunk 4]
```diff
@@ -835,6 +835,78 @@ class LatentState:
...
+        self.patch_token = patch_token
+
+        self.pred_blocks = nn.ModuleList()
+
+        # first map to intermediate_dim to introduce a bottleneck
+        self.pred_blocks.append(nn.Linear(in_dim, intermediate_dim, bias=False))
+
+        for _ in range(self.cf.sslpred_num_blocks):
+            self.pred_blocks.append(
+                MultiSelfAttentionHead(
```
Comment (clessig) [inline src/weathergen/model/engines.py:863]: We should have a transformer_block module for attention + MLP (can/should be in a separate PR)

File: src/weathergen/model/model.py [hunk 5]
```diff
@@ -445,10 +446,12 @@ def create(self) -> "Model":
                         patch_token=True,
                     )
                 elif loss == "JEPA":
-                    self.latent_heads[loss] = LatentPredictionHead(
+                    self.latent_heads[loss] = TransformerPredictionHead(
```
Comment (clessig) [inline src/weathergen/model/model.py:449]: Why is here no if-statement to chose between LatentPredictionHead and TransformerPredictionHead? LatentPredictionHead should be renamed to LatentPredictionHeadMLP (and then I would also prefer LatentPredictionHeadTransformer).

File: config/config_physical_jepa.yml (hunk 3 above)
Comment (clessig) [reply config/config_physical_jepa.yml:68]: Option A is model but it's ultimately something specific to the JEPA loss term.

**Comment by sophie-xhonneux** [comment]
I clean up the PR a bit, but didn't know how to open a PR against this PR so I put it here https://github.com/ecmwf/WeatherGenerator/pull/1649

**Comment by clessig** [comment]
> I clean up the PR a bit, but didn't know how to open a PR against this PR so I put it here #1649

You need open the PR against the repo Kerem is using (MeteoSwiss fork of WeatherGenerator).
